File: js/22_05_18/stack.js
var stack // data structure array
var front = 0;
var back = 0; // data index cursor


const stackInit = (n) => {
  stack = new Array(n);
  front = 0;
  back = stack.length - 1;
  console.log(stack.length)
}

const push = (InputNumber) => {
  if (stack[front] !== undefined){
    console.log("stack is full")
    return;
  }
  stack[front] = InputNumber;
  front += 1;
  if (front === stack.length) {
    front = 0;
  }
}

const pop = () => {
  if (stack[back] === undefined){
    console.log("stack is empty")
    return;
  }
  stack[back] = undefined;
  back -= 1;
}

const printAll = () => {
  console.log(stack)
}

// Run Test
stackInit(3)
push(1)
push(2)
push(3)
push(5)
pop()
pop()
pop()
printAll()

File: go/stack/stack.go
package stack

import "fmt"

type Stack struct {
	array []int
	sp    int
}

func (s *Stack) StackInit(n int) []int {
	s.array = make([]int, n)
	s.sp = 0
	fmt.Println(s.array)
	return s.array
}

func (s *Stack) Push(inputNumber int) []int {
	c := s.sp
	if c >= len(s.array) {
		fmt.Println("stack is full")
	} else {
		s.array[c] = inputNumber
	}
	s.PushCursor()
	fmt.Println("inputNumber", inputNumber)
	return s.array
}

func (s *Stack) PushCursor() int {
	if len(s.array) == s.sp {
		fmt.Println("stack is full")
	} else {
		s.sp++
	}
	return s.sp
}

func (s *Stack) PopCursor() int {
	if s.sp == 0 {
		fmt.Println("stack is empty")
	} else {
		s.sp--
	}
	return s.sp
}

func (s *Stack) Pop() []int {
	c := s.sp - 1
	s.array[c] = 0
	s.PopCursor()
	fmt.Println("pop", c)
	return s.array
}

func main() {
	var s Stack
	s.StackInit(3)
	fmt.Println(s.Push(3))
	fmt.Println(s.Push(4))
	fmt.Println(s.Push(5))
	fmt.Println(s.Push(4))
	fmt.Println(s.Pop())
	fmt.Println(s.Pop())
	fmt.Println(s.Pop())
	fmt.Println(s.Pop())
}
